flow/shared/alerting/alerting.go

```diff
@@ -0,0 +1,96 @@
+package alerting
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"log/slog"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+// alerting service, no cool name :(
+type Alerter struct {
+	catalogPool *pgxpool.Pool
+	logger      *slog.Logger
+}
+
+func registerSendersFromPool(catalogPool *pgxpool.Pool) ([]*slackAlertSender, error) {
+	rows, err := catalogPool.Query(context.Background(),
+		"SELECT service_type,service_config FROM peerdb_stats.alerting_config")
+	if err != nil {
+		return nil, fmt.Errorf("failed to read alerter config from catalog: %w", err)
+	}
+
+	var slackAlertSenders []*slackAlertSender
+	var serviceType, serviceConfig string
+	_, err = pgx.ForEachRow(rows, []any{&serviceType, &serviceConfig}, func() error {
+		switch serviceType {
+		case "slack":
+			var slackServiceConfig slackAlertConfig
+			err = json.Unmarshal([]byte(serviceConfig), &slackServiceConfig)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal Slack service config: %w", err)
+			}
+
+			slackAlertSenders = append(slackAlertSenders, newSlackAlertSender(&slackServiceConfig))
+		default:
+			return fmt.Errorf("unknown service type: %s", serviceType)
+		}
+		return nil
+	})
+
+	return slackAlertSenders, nil
+}
+
+// doesn't take care of closing pool, needs to be done externally.
+func NewAlerter(catalogPool *pgxpool.Pool) *Alerter {
+	return &Alerter{
+		catalogPool: catalogPool,
+		logger:      slog.Default(),
+	}
+}
+
+// Only raises an alert if another alert with the same key hasn't been raised
+// in the past 15 minutes
+func (a *Alerter) AlertIf(ctx context.Context, alertKey string, alertMessage string) {
+	if a.catalogPool != nil {
```

Comment:
when can it be nil?

Reply:
shouldn't be, but writing it defensively